### DISCUSSION TITLE
Creare fattura da Google Sheet

### DISCUSSION THREAD

**Discussion by KangiSport** [fattureincloud/api#480] [Q&A] [answered]
Ciao a tutti, ho creato l'applicazione su Fattureincloud per ottenere il token personale al quale ho dato tutti i permessi possibili e ho recuperato il company ID.
Ho creato uno script su Apps Script di Google collegato a un google sheet in cui ci sono i dati per creare la fattura.

Il foglio è strutturato così:

![foglio](https://github.com/user-attachments/assets/fff525ed-a050-4941-a2e6-bc5cc036f4ec)

A seconda del numero in H1 lo script prende in considerazione gli ordini relativi a quel mese ed ognuno di essi diventa un item della fattura.

Lo script è il seguente (sostituisco i dati sensibili con asterischi):

```
function creaFatturaFattureincloud() {
  const sheet = SpreadsheetApp.getActiveSpreadsheet().getSheetByName("FATTURE");
  const meseDaFatturare = sheet.getRange("J1").getValue();
  const meseInLettere = getMonthInWords(meseDaFatturare);

  const token = "*******"; // Inserisci il tuo token personale qui
  const companyId = "*******";  // Inserisci il tuo company_id qui
  const url = `https://api-v2.fattureincloud.it/c/${companyId}/invoices`;

  const datiFattura = {
    type: "invoice", // Tipo di documento
    entity: {
      name: "*******",
      vat_number: "********",
      tax_code: "********",
      address: "*********",
      postal_code: "*******",
      city: "*****",
      province: "******",
      country: "*****",
      email: ********"
    },
    date: new Date().toISOString().split('T')[0], // Data corrente
    visible_subject: `Servizio SUBLIREADY - ${meseInLettere}`,
    currency: {
      id: "EUR"
    },
    items_list: []
  };

  // Legge i dati e filtra per il mese specificato
  const rows = sheet.getDataRange().getValues();
  for (let i = 1; i < rows.length; i++) {
    const mese = rows[i][0];
    const ordine = rows[i][2];
    const totale = rows[i][5];
    
    if (mese === meseDaFatturare) {
      datiFattura.items_list.push({
        name: `Ordine ${ordine}`,
        quantity: 1,
        net_price: totale,
        vat: {
          id: "22" // IVA al 22%
        }
      });
    }
  }

  const options = {
    method: "post",
    headers: {
      "Authorization": "Bearer " + token,
      "Content-Type": "application/json"
    },
    payload: JSON.stringify(datiFattura),
    muteHttpExceptions: true
  };

  const response = UrlFetchApp.fetch(url, options);
  const result = JSON.parse(response.getContentText());

  if (response.getResponseCode() === 200) {
    Logger.log("Fattura creata con successo!");
  } else {
    Logger.log("Errore nella creazione della fattura: " + response.getContentText());
  }
}

// Funzione per convertire il numero del mese in lettere
function getMonthInWords(mese) {
  const mesi = ["Gennaio", "Febbraio", "Marzo", "Aprile", "Maggio", "Giugno", "Luglio", "Agosto", "Settembre", "Ottobre", "Novembre", "Dicembre"];
  return mesi[mese - 1];
}
```

Provo ad eseguire lo script e mi da sempre errore:

Errore nella creazione della fattura: {"error":{"message":"","code":0}}

Ho fatto un test con uno script più semplice anche su Postman, ma anche lì mi da errore (ho coperto il company ID):

![postman](https://github.com/user-attachments/assets/ac8956bc-ad1f-4a44-97ab-e59a48db5c51)

Il codice errore è così generico che nemmeno chatGPT, che mi ha seguito nel fare lo script, sa a cosa si riferisce.

Questo è il log nella scheda dell'applicazione in fattureincloud:

![log](https://github.com/user-attachments/assets/e786e68e-ae60-4516-865f-48a162832570)

Qualcuno può aiutarmi?
Grazie.

**Comment by valmoz** [comment, marked as answer]
Buongiorno,
l'endpoint che sta chiamando non esiste. Deve utilizzare il metodo [Create Issued Document](https://developers.fattureincloud.it/api-reference#post-/c/-company_id-/issued_documents)

Le consiglio di seguire le guide seguenti:
- [Google Sheets Apps Script Quickstart](https://developers.fattureincloud.it/docs/quickstarts/apps-script-quickstart)
- [Invoice Creation](https://developers.fattureincloud.it/docs/guides/invoice-creation)

In generale, ChatGPT funziona bene solo con Librerie utilizzate ampiamente a livello mondiale (per poter avere materiale per addestrare il modello), mentre per la casistica "AppsScript + Fatture in Cloud APIs" immagino non ci siano molti esempi online, per questo si verificano queste allucinazioni nel tentativo di "indovinare" la risposta più probabile.

In teoria, può anche provare ad importare la nostra [SDK JS](https://developers.fattureincloud.it/docs/sdks/javascript-sdk) come libreria esterna in Apps Script (come descritto [qui](https://developers.google.com/apps-script/guides/libraries))

A disposizione qualora avesse ulteriori domande.

Grazie
Mauro

**Reply by valmoz** [reply]
Buongiorno @KangiSport è riuscito a risolvere?

**Reply by KangiSport** [reply]
Buongiorno, mi scusi avrei voluto risponderle prima ma non sono riuscito. Intanto grazie per essersi preso il tempo di rispondermi. Non ho importato l'SDK JS perché sarebbe stato uno sforzo eccessivo vista la particolarità e modestia dello script che mi serviva ma ho dato un'occhiata ai link che mi ha passato e sono stati molto utili tanto che infine sono riuscito ad ottenere quel che volevo. Ho penato abbastanza anche per risolvere piccoli problemi come l'incongruenza dei totali ottenuti in tabella excel e quelli attesi da fattureincloud per una mera questione di approsimazione dei decimali.. ma infine ce l'ho fatta! Dovessi aver bisogno di risolvere altri script potrei chiedere ancora il suo aiuto! Grazie di nuovo!

**Reply by valmoz** [reply]
Perfetto, sono lieto di essere stato utile.
Buona giornata
Mauro
